### PR TITLE
Fix `Number#step`

### PR DESCRIPTION
**Pull request by straight-shoota**
`Number#step` was broken for example around definition ranges. This is a complete re-implementation with an extensive spec suite.

### CODE REVIEW

**Comment by straight-shoota** [comment]
Would be great to get a review on this as it's holding up consecutive work (and is fixing bugs on its own).

File: src/number.cr [hunk 1]
```diff
@@ -87,50 +87,138 @@ struct Number
...
+  #
+  # In those cases the iteration is empty.
+  def step(*, to limit = nil, by step)
+    # type of current should be the result of adding `step`:
+    current = self + (step - step)
+
+    if limit == current
+      yield current
+      return
+    elsif step.zero?
```
Comment (asterite) [inline src/number.cr]: It seems Ruby allows a step of zero. Maybe it's fine?

File: src/number.cr [hunk 2]
```diff
@@ -87,50 +87,138 @@ struct Number
...
+    end
+
+    direction = step.sign
+
+    if limit
+      return unless (limit <=> current).try(&.sign) == direction
+
+      yield current
+
+      while ((limit - current) <=> step) == direction
```
Comment (asterite) [inline src/number.cr]: Do you have benchmarks to see if these space ship comparisons didn't affect the performance?

File: src/number.cr (hunk 1 above)
Comment (asterite) [reply src/number.cr]: That said, we can loos3n that restriction later on (Ruby did that when they added the keyword arguments alternative)

Comment (straight-shoota) [reply src/number.cr]: Zero step size typically results in an infinite loop, so it's easy to shoot yourself in the foot with that. 
If there's a valid use case for this, I'm fine with chaning it.
But I'd prefer the method to raise when I give it silly input.

**Review by asterite — APPROVED** [review]
Looks great! Feel free to merge. We can always improve the performance (if it changed) later on, and loosen the zero step restriction.

File: src/number.cr (hunk 2 above)
Comment (straight-shoota) [reply src/number.cr]: I didn't check any benchmark and solely focused on correctness 👍 Optimization can follow.

**Pull request merged**